src/MoveUntilTouchTopicController.cpp

```diff
@@ -0,0 +1,216 @@
+  tareActionClient = std::unique_ptr<TareActionClient>(new TareActionClient(nh, ft_tare_name));
+
+  // start action server
+  mFTThresholdActionServer.reset(
+      new actionlib::ActionServer<SetFTThresholdAction>{
+          nh, "set_forcetorque_threshold",
+          std::bind(&MoveUntilTouchTopicController::setForceTorqueThreshold, this,
+                    std::placeholders::_1),
+          false});
+  mFTThresholdActionServer->start();
+
+  // initialize base trajectory controller
+  return initController(robot, nh);
+}
+
+//=============================================================================
+void MoveUntilTouchTopicController::forceTorqueDataCallback(const geometry_msgs::WrenchStamped& msg)
+{
+  mForce.x() = msg.wrench.force.x;
+  mForce.y() = msg.wrench.force.y;
+  mForce.z() = msg.wrench.force.z;
+  mTorque.x() = msg.wrench.torque.z;
+  mTorque.y() = msg.wrench.torque.z;
+  mTorque.z() = msg.wrench.torque.z;
+}
+
+//=============================================================================
+void MoveUntilTouchTopicController::taringTransitionCallback(TareActionClient::GoalHandle goalHandle) {
+  if (goalHandle.getResult()) {
+    if (goalHandle.getResult()->success) {
+      ROS_INFO("Taring completed!");
+      mTaringCompleted.store(true);
+    }
+  }
+}
+
+//=============================================================================
+void MoveUntilTouchTopicController::starting(const ros::Time& time)
+{
+  // start asynchronous tare request
+  ROS_INFO("Starting Taring");
+  pr_control_msgs::TriggerGoal goal;
+  mTareGoalHandle = tareActionClient->sendGoal(goal, boost::bind(&MoveUntilTouchTopicController::taringTransitionCallback, this, _1));
+
+  // start base trajectory controller
+  startController(time);
+}
+
+//=============================================================================
+void MoveUntilTouchTopicController::stopping(const ros::Time& time)
+{
+  // stop base trajectory controller
+  stopController(time);
+}
+
+//=============================================================================
+void MoveUntilTouchTopicController::update(const ros::Time& time,
+                                      const ros::Duration& period)
+{
+  // update base trajectory controller
+  updateStep(time, period);
+}
+
+//=============================================================================
+bool MoveUntilTouchTopicController::shouldAcceptRequests() { return isRunning(); }
+
+//=============================================================================
+bool MoveUntilTouchTopicController::shouldStopExecution(std::string& reason)
+{
+  // inelegent to just terminate any running trajectory,
+  // but we must guarantee taring completes before starting
+  if (!mTaringCompleted.load()) {
+    ROS_WARN("taring not yet completed!");
+    return true;
+  }
+
+  double forceThreshold = mForceThreshold.load();
+  double torqueThreshold = mTorqueThreshold.load();
+
+  forceTorqueDataMutex.lock();
```

Comment:
I think we generally prefer to use [`lock_guard`](https://en.cppreference.com/w/cpp/thread/lock_guard) or [`unique_lock`](https://en.cppreference.com/w/cpp/thread/unique_lock) over explicit mutex lock/unlock? @jslee02 would know more.

Reply:
Yes, there is no technically difference between them, but we conventionally prefer using `lock_guard`, which is RAII wrapper, (`lock_guard` seems to suffice for our use case; see this [post](https://stackoverflow.com/a/20516876/3122234)) over manually locking and unlocking.